src/Microsoft.ML.Data/Scorers/FeatureContributionCalculatingTransformer.cs

```diff
@@ -41,88 +49,192 @@ namespace Microsoft.ML.Runtime.Data
+                else
+                    result.Add(new Schema.DetachedColumn(DefaultColumnNames.FeatureContributions, new VectorType(NumberType.R4, _roleMappedSchema.Feature.Type.AsVector), builder.GetMetadata()));
+
+                return result.ToArray();
+            }
+
+            protected Delegate GetScoreGetter(Row input, int iinfo, Func<int, bool> active)
+            {
+                var genericRow = _genericRowMapper.GetRow(input, col => active(col));
+                return RowCursorUtils.GetGetterAsDelegate(genericRow, iinfo);
+            }
+
+            protected Delegate GetFeatureContributionGetter(Row input, int iinfo, Func<int, bool> active)
+            {
+                if (active(iinfo))
+                {
+                    return _bindableMapper.Stringify
+                        ? _bindableMapper.GetTextContributionGetter(input, _roleMappedSchema.Feature.Index, _slotNames)
+                        : _bindableMapper.GetContributionGetter(input, _roleMappedSchema.Feature.Index);
+                }
+                return null;
+            }
+
+            protected override Delegate MakeGetter(Row input, int iinfo, Func<int, bool> active, out Action disposer)
+            {
+                disposer = null;
+                if (iinfo < _outputGenericSchema.ColumnCount)
+                    return GetScoreGetter(input, iinfo, active);
+                else
+                    return GetFeatureContributionGetter(input, iinfo, active);
+            }
         }
 
-        private sealed class BindableMapper : ISchemaBindableMapper, ICanSaveModel, IPredictor
+        // TODO documentation
+        internal sealed class BindableMapper : ISchemaBindableMapper, ICanSaveModel, IPredictor
         {
+            public sealed class Arguments : ScorerArgumentsBase
+            {
+                [Argument(ArgumentType.AtMostOnce, HelpText = "Number of top contributions", SortOrder = 1)]
+                public int Top = FeatureContributionCalculatingEstimator.Defaults.Top;
+
+                [Argument(ArgumentType.AtMostOnce, HelpText = "Number of bottom contributions", SortOrder = 2)]
+                public int Bottom = FeatureContributionCalculatingEstimator.Defaults.Bottom;
+
+                [Argument(ArgumentType.AtMostOnce, HelpText = "Whether or not output of Features contribution should be normalized", ShortName = "norm", SortOrder = 3)]
+                public bool Normalize = FeatureContributionCalculatingEstimator.Defaults.Normalize;
+
+                [Argument(ArgumentType.AtMostOnce, HelpText = "Whether or not output of Features contribution in string key-value format", ShortName = "str", SortOrder = 4)]
+                public bool Stringify = FeatureContributionCalculatingEstimator.Defaults.Stringify;
+
+                // REVIEW: the scorer currently ignores the 'suffix' argument from the base class. It should respect it.
```

Comment:
Is this comment still valid?

Reply:
Yes it is, although I don't think we should keep the Arguments object in the constructor as we don't have many arguments for this estimator and we should be able to specify them directly. I have not seen the suffix setting in any other estimator also.

---
In reply to: [239966486](https://github.com/dotnet/machinelearning/pull/1847#discussion_r239966486) [](ancestors = 239966486)